### ISSUE TITLE
Question About the memory

### ISSUE THREAD

**Issue by mrpre** [RedisGears/RedisGears#389]
I find each time I execute `RG.PYEXECUTE "gb = GearsBuilder();gb.run()"`, the `CurrAllocated` showed in `rg.pystats` increase.

```shell
127.0.0.1:6379> rg.pystats
1) "TotalAllocated"
2) (integer) 10585184
3) "PeakAllocated"
4) (integer) 4045354
5) "CurrAllocated"
6) (integer) 3727096
127.0.0.1:6379> RG.PYEXECUTE "gb = GearsBuilder();gb.run()"
1) (empty array)
2) (empty array)
127.0.0.1:6379> RG.PYEXECUTE "gb = GearsBuilder();gb.run()"
1) (empty array)
2) (empty array)
127.0.0.1:6379> rg.pystats
1) "TotalAllocated"
2) (integer) 10701874
3) "PeakAllocated"
4) (integer) 4045354
5) "CurrAllocated"
6) (integer) 3736872
```
After running  it hundreds of times, the CurrAllocated becomes 4854904.  When the memory is freed ?

**Comment by MeirShpilraien** [comment]
Hey @mrpre 

The CurrAllocated is in bytes so 4854904 is only around 5MB which is reasonable. The python interpreter keeps some commonly used objects in memory to reuse them (like lists, sets, even numbers). So seen memory increase in such levels is making sense. If you see the python memory gets to GB and not getting down then it's strange and should be investigated (might be a leak or a registration that saves a huge global collection...)

**Comment by mrpre** [comment]
@MeirShpilraien 

I run the command lot of times and find the `CurrAllocated ` becomes 659000562 bytes (659MB)
```shell
127.0.0.1:6379> rg.pystats
1) "TotalAllocated"
2) (integer) 7627530778
3) "PeakAllocated"
4) (integer) 659048816
5) "CurrAllocated"
6) (integer) 659000562
127.0.0.1:6379>
```

Also I find `PyFlatExecution_Destruct` isn't called after the command returns.
I try to figure out the reference of gb by using `import sys;gb = GearsBuilder();print(sys.getrefcount(gb));gb.run();print(sys.getrefcount(gb))`  and find both result are 2.  
It means the the reference of `gb` is 1 and should be gc.

**Comment by MeirShpilraien** [comment]
amm ok thanks for the analysis, I will try to take a look. Can you also try run it like this and let me know if you see the same issue?
```
GearsBuilder().run()
```

I guess because you put it in a variable it saves in the global dictionary.

Also, can you specify the RedisGears version you run with?

**Comment by mrpre** [comment]
@MeirShpilraien you are right when running `GearsBuilder().run()` the memory usage doesn't increase.
Does it mean that global variables are not allowed to be used in RedisGears ?

**Comment by MeirShpilraien** [comment]
@mrpre not at all, they are totally allowed to be used and its definitely a bug. I just wanted to do this test to point me in the right direction. I will look into it.

**Comment by MeirShpilraien** [comment]
Can you just specify the version you run with?

**Comment by mrpre** [comment]
I didn't use the release version.  

Redis   
branch unstable  
commitid 5977a94842a25140520297fe4bfda15e0e4de711    


RedisGears 
branch master  
commitid ced15413f0bc46e7a93f40be4d4b1809e8e681e6

**Comment by MeirShpilraien** [comment]
@mrpre it looks like it happened only if you put the `GearsBuilder` in a global variable (other global variables are fine). Looks like there is some cycling reference that prevents it from being free. As a workaround, until we fix it, just make sure not to put it in a global variable (or set the `gc` to `None` before the code finishes).
